mysqltest/mysqltest.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package testserver provides an ephemeral MySQL server for testing database integration. It's
+// designed to be used in code that needs to work inside and outside Google.
+package mysqltest
+
+import (
+	"log"
+	"os"
+	"testing"
+)
+
+var connInfo ConnInfo
+
+// A Stopper stops the Docker container that is started by this package.
+type Stopper func() error
+
+// ConnInfo specifies how to connect to the server that is created by this package.
+type ConnInfo struct {
+	Username string
+	Password string
+	Hostname string
+	Port     int
+}
+
+// Get returns the address and credentials of the MySQL server. The server must already have been
+// initialized by calling TestMainHelper(), or this function will panic.
+func Get() ConnInfo {
```

Comment:
I don't really like this global object. What's the use case for doing this instead of returning the config from `startContainer`?

Reply:
This is interesting design question. I chose this route to minimize boilerplate inside the tests who import this code.

Here's how the user uses this library:

```
func TestMain(m *testing.M) {
	mysqltest.TestMainHelper(m)
}

func TestFoo(t *testing.T) {
	connInfo := mysqltest.Get()
	doSomethingWith(connInfo)
	...
}
```

This is good because:

 - There's no plumbing of values from TestMain to TestFoo
 - TestMain is only one line
 - It's impossible to do it wrong. This is wonderful thing in API design. There's no `defer` to forget.

This seems somewhat preferable to the alternative approach from EN:

```
var testDatabaseInstance *database.TestInstance

func TestMain(m *testing.M) {
  testDatabaseInstance = database.MustTestInstance()
  defer testDatabaseInstance.MustClose()
  m.Run()
}
```

Reply:
The current approach limits to exactly one instance per package, and I think we're going to find situations where we need multiple independent database instances within a single package. I like the idea of being "impossible to do it wrong", but I don't think that tradeoff is worth the flexibility of having users create a pool.

Reply:
OK, I can live with that. Would you be OK with offering both APIs?

One API would be the `TestMainHelper` API that's already here, which has the benefit of removing boilerplate plumbing when there are lots of database tests. The other API would be a Start() method that returns a Closer, similar to EN.

Reply:
I prefer only one API (do one thing and do it well philosophy)

Reply:
OK, done.